README.md
<p align="center"><a href="https://github.com/crazy-max/firefox-history-merger" target="_blank"><img width="100"src="https://raw.githubusercontent.com/crazy-max/firefox-history-merger/master/.res/firefox-history-merger.png"></a></p>

<p align="center">
  <a href="https://github.com/crazy-max/firefox-history-merger/releases/latest"><img src="https://img.shields.io/github/release/crazy-max/firefox-history-merger.svg?style=flat-square" alt="GitHub release"></a>
  <a href="https://github.com/crazy-max/firefox-history-merger/releases/latest"><img src="https://img.shields.io/github/downloads/crazy-max/firefox-history-merger/total.svg?style=flat-square" alt="Total downloads"></a>
  <a href="https://github.com/crazy-max/firefox-history-merger/actions?workflow=build"><img src="https://img.shields.io/github/workflow/status/crazy-max/firefox-history-merger/build?label=build&logo=github&style=flat-square" alt="Build Status"></a>
  <a href="https://goreportcard.com/report/github.com/crazy-max/firefox-history-merger"><img src="https://goreportcard.com/badge/github.com/crazy-max/firefox-history-merger?style=flat-square" alt="Go Report"></a>
  <a href="https://www.codacy.com/app/crazy-max/firefox-history-merger"><img src="https://img.shields.io/codacy/grade/85a6dc4ddaf14aeba2c4f1876126d785.svg?style=flat-square" alt="Code Quality"></a>
  <br /><a href="https://github.com/sponsors/crazy-max"><img src="https://img.shields.io/badge/sponsor-crazy--max-181717.svg?logo=github&style=flat-square" alt="Become a sponsor"></a>
  <a href="https://www.paypal.me/crazyws"><img src="https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square" alt="Donate Paypal"></a>
</p>

## About

**firefox-history-merger** is a CLI application written in [Go](https://golang.org/) to merge history of [Firefox](https://www.mozilla.org/en-US/firefox/) and repair missing favicons with ease.

![](.res/screenshot.png)
> Screenshot of firefox-history-merger

___

* [Features](#features)
* [Requirements](#requirements)
* [Download](#download)
* [Usage](#usage)
  * [Info](#info)
  * [Merge](#merge)
  * [Repair favicons](#repair-favicons)
* [About Firefox places](#about-firefox-places)
  * [Places expiration limit](#places-expiration-limit)
  * [Databases schema](#databases-schema)
* [How can I help?](#how-can-i-help)
* [License](#license)

## Features

* Merge history (`moz_places`) from `places.sqlite` file
* `moz_historyvisits` are also merged
* [Frecency](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Places/Frecency_algorithm) is recalculated during the merge
* Repair favicons
* Display info about `places.sqlite`
* Optimize database into a minimal amount of disk space

## Requirements

Schema version that can be used is **v39 (Firefox 57)** to **v53 (Firefox 81)**. Check if your `places.sqlite` is compatible with `info` command.

If your database is not compatible, you can copy your `places.sqlite` in a compatible Firefox [profile folder](https://support.mozilla.org/en-US/kb/profiles-where-firefox-stores-user-data), then launch/close Firefox and your `places.sqlite` should now be compatible. 

## Download

You can download the application matching your platform on the [**releases page**](https://github.com/crazy-max/firefox-history-merger/releases/latest).

## Usage

First close Firefox and copy `places.sqlite` and `favicons.sqlite` files from [your Firefox profile folder](https://support.mozilla.org/en-US/kb/profiles-where-firefox-stores-user-data) somewhere.

In the following examples, this folder structure will be used :

```
favicons.sqlite
firefox-history-merger.exe
places.sqlite
places_to_merge.sqlite
```

### Info

You probably want to have more info about your `places.sqlite` database you have before merging history:

```
$ firefox-history-merger info places.sqlite
Sun, 23 Jun 2019 03:43:47 CEST INF Schema version:         v52 (Firefox >= 65)
Sun, 23 Jun 2019 03:43:48 CEST INF Compatible:             true
Sun, 23 Jun 2019 03:43:48 CEST INF Places entries:         1934
Sun, 23 Jun 2019 03:43:48 CEST INF Historyvisits entries:  1678
Sun, 23 Jun 2019 03:43:48 CEST INF Last used on:           2019-06-21 02:31:46
```

### Merge

Now execute this command to merge `places_to_merge.sqlite` with your working `places.sqlite` :

```
$ firefox-history-merger merge places.sqlite places_to_merge.sqlite
Sun, 23 Jun 2019 04:33:58 CEST INF 5199 places will be merged
Sun, 23 Jun 2019 04:33:58 CEST INF Merging 1000 places (1/6)...
Sun, 23 Jun 2019 04:33:59 CEST INF Merging 1000 places (2/6)...
Sun, 23 Jun 2019 04:34:01 CEST INF Merging 1000 places (3/6)...
Sun, 23 Jun 2019 04:34:03 CEST INF Merging 1000 places (4/6)...
Sun, 23 Jun 2019 04:34:05 CEST INF Merging 1000 places (5/6)...
Sun, 23 Jun 2019 04:34:07 CEST INF Merging 199 places (6/6)...
Sun, 23 Jun 2019 04:34:09 CEST INF Optimizing database...
Sun, 23 Jun 2019 04:34:09 CEST INF Finished created=4478 errors=0 total=5199 updated=721
```

### Repair favicons

If you want you can also repair missing favicons. It can take a long time depending on the number of entries in your history, since the favicon will be recovered for each entry if there is no attached favicon.

```
$ firefox-history-merger repair-favicons places.sqlite favicons.sqlite
Sun, 23 Jun 2019 04:29:02 CEST INF Checking 1948 places...
Sun, 23 Jun 2019 04:29:02 CEST INF Favicon repaired favicon_id=18575 left=1938 places_id=33864 url=https://github.com/
Sun, 23 Jun 2019 04:29:05 CEST ERR Cannot get favicon error="no favicon found" favicon_id=0 left=1845 places_id=200497 url=http://www.videojs.com/
...
Sun, 23 Jun 2019 04:32:54 CEST ERR Cannot get favicon error="no favicon found" favicon_id=0 left=1 places_id=1955773 url=https://docs.docker.com/v17.09/engine/admin/logging/view_container_logs/
Sun, 23 Jun 2019 04:32:54 CEST INF Finished errors=618 linked=873 repaired=402 skipped=23 total=1948 valid=32
```

## About Firefox places

### Places expiration limit

As stipulated in this [StackExchange thread](https://superuser.com/questions/895302/how-do-i-set-max-browsing-history-size/995459#995459), the relevant function is `_loadPrefs()` in [nsPlacesExpiration.js](https://dxr.mozilla.org/mozilla-central/source/toolkit/components/places/nsPlacesExpiration.js#714). It tries to determine the limit by first checking `PREF_MAX_URIS` (which is `"max_pages"`). If `PREF_MAX_URIS` isn't set or is a negative value, it then attempts to calculate the 'optimal database size' based on available memory and disk space.

So if you want to retain all history ('infinite' history), you could set `places.history.expiration.max_pages` to 2147483647 (the highest 32-bit signed integer) in `about:config`.

Once you create it and set it, check the value of `places.history.expiration.transient_current_max_pages`. It should automatically adjust itself to match your `max_pages` setting. 

### Databases schema

Database schema version is stored in `user_version` [pragma statement](https://sqlite.org/pragma.html). Linked Firefox version to database schema can be found in [Database.cpp](https://dxr.mozilla.org/mozilla-central/source/toolkit/components/places/Database.cpp#993).

![Firefox v53 database schema](.res/schemas/v53.png)
> [Edit this diagram](https://www.draw.io/?title=firefox_v53.png&url=https%3A%2F%2Fraw.githubusercontent.com%2Fcrazy-max%2Ffirefox-history-merger%2Fmaster%2F.res%2Fschemas%2Fv53.png%3Ft%3D0) in your browser. 

## How can I help?

All kinds of contributions are welcome :raised_hands:! The most basic way to show your support is to star :star2: the project, or to raise issues :speech_balloon: You can also support this project by [**becoming a sponsor on GitHub**](https://github.com/sponsors/crazy-max) :clap: or by making a [Paypal donation](https://www.paypal.me/crazyws) to ensure this journey continues indefinitely! :rocket:

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.<br />
Icon credit to [Zlatko Najdenovski](http://pixelbazaar.com/) (firefox icon) and [BomSymbols](https://creativemarket.com/BomSymbols) (clock icon).


CHANGELOG.md
# Changelog

## 2.12.0 (2020/10/22)

* Same schema (v53) for Firefox 79, 80, 81

## 2.11.0 (2020/10/22)

* Same schema (v53) for Firefox 78

## 2.10.0 (2020/06/27)

* Same schema (v53) for Firefox 77

## 2.9.0 (2020/06/27)

* Same schema (v53) for Firefox 76
* Update deps

## 2.8.0 (2020/04/17)

* Same schema (v53) for Firefox 75

## 2.7.0 (2020/04/17)

* Same schema (v53) for Firefox 74

## 2.6.0 (2020/03/05)

* Same schema (v53) for Firefox 73
* Update deps

## 2.5.0 (2020/01/29)

* Same schema (v53) for Firefox 72

## 2.4.0 (2019/12/15)

* Same schema (v53) for Firefox 71
* Go 1.13

## 2.3.0 (2019/11/03)

* Same schema (v53) for Firefox 70

## 2.2.2 (2019/09/19)

* Use softprops/action-gh-release to create GitHub release
* Switch to [ghaction-xgo](https://github.com/crazy-max/ghaction-xgo) GitHub Action

## 2.2.0 (2019/09/15)

* Add implementation for Firefox 69 schema version v53
* Switch To GitHub Actions

## 2.1.1 (2019/07/09)

* Do not set origin id

## 2.1.0 (2019/07/09)

* Same schema (v52) for Firefox 68
* Fix `favicon_id` column does not exist (Issue #10 #17 #18)

## 2.0.2 (2019/06/23)

* Small typo
* Coding style

## 2.0.1 (2019/06/23)

* Same schema (v52) for Firefox 66 and 67

## 2.0.0 (2019/06/23)

* Minimal schema version v39 (Firefox 57) required
* Paginate and use transactions while merging places to improve performance
* Improve favicons repairing performance
* Handle os signals
* Better error handling
* Switch to zerolog as logger
* Remove unecessary modules
* Review project structure
* Go 1.12

## 1.65.0 (2019/02/15)

* Same schema (v52) for Firefox 65
* Review build process
* Switch to TravisCI (com)

## 1.64.1 (2018/12/22)

* Add `--limit` flag
* Add infos about places databases found

## 1.64.0 (2018/12/14)

* Same schema (v52) for Firefox 64

## 1.63.0 (2018/10/27)

* Same schema (v52) for Firefox 63 (Issue #9)

## 1.62.0 (2018/10/27)

* Add implementation for Firefox 62 schema version v52 (Issue #8)
  * `moz_hosts` renamed `moz_origins` and column `typed` removed
  * Index `origin_id` added to `moz_places`
* Wrong last used date
* More intuitive variables names
* Upgrade to Go 1.11
* Use [go mod](https://golang.org/cmd/go/#hdr-Module_maintenance) instead of `dep`

## 1.61.0 (2018/07/05)

* Add implementation for Firefox 61 schema version v47 (Issue #6)

## 1.60.0 (2018/05/11)

* Add implementation for Firefox 60 schema version v43

## 1.59.0 (2018/03/15)

* Same schema (v41) for Firefox 59 (Issue #4)

## 1.58.0 (2018/01/27)

* Add implementation for Firefox 58 schema version v41

## 1.57.2 (2018/01/27)

* Mozilla has moved the old `moz_favicons` table to the `favicons.sqlite` db
* Add command `optimize` to optimize a database into a minimal amount of disk space
* Make icon cache optional and use flag `--enable-cache` to enable it (Issue #2)
* Goroutine can cause a database lockup on Linux
* Put more explicit messages when checking arguments (Issue #1)
* Check database compatibility using schema version (disable auto migration)
* Base semantic versioning on Firefox version
* Prepare implementation for Firefox 58
* New logger
* Update libs

## 0.1.1 (2017/12/20)

* Create artifacts for Linux and macOS

## 0.1.0 (2017/12/10)

* Initial version


cmd/merge.go
package cmd

import (
	"os"
	"strconv"
	"strings"

	"github.com/crazy-max/firefox-history-merger/sqlite"
	"github.com/crazy-max/firefox-history-merger/sqlite/places"
	. "github.com/crazy-max/firefox-history-merger/utils"
	"github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
	"gopkg.in/cheggaaa/pb.v1"
)

type MergeResult struct {
	Created int
	Updated int
	Errors  int
}

var (
	mergeCmd = &cobra.Command{
		Use:     "merge",
		Short:   "Merge a working places.sqlite with other ones",
		Example: AppName + ` merge "/home/user/places.sqlite" "/home/user/folder/containing/places_sqlite/"`,
		Run:     mergeRun,
	}

	doMergeFull             bool
	doMergeMinimal          bool
	doMergeAddOrigins       bool
	doMergeAddHistoryvisits bool
	limit                   int

	lastID struct {
		MozPlaces        int
		MozHistoryvisits int
		MozOrigins       int
	}

	mergeLogFile *os.File
	mergeLog     = logrus.New()
)

func init() {
	mergeCmd.PersistentFlags().BoolVar(&doMergeMinimal, "merge-minimal", true, "Merge moz_places")
	mergeCmd.PersistentFlags().BoolVar(&doMergeFull, "merge-full", false, "Merge moz_places, moz_origins and moz_historyvisits")
	mergeCmd.PersistentFlags().BoolVar(&doMergeAddHistoryvisits, "merge-add-historyvisits", false, "Add moz_historyvisits to merge")
	mergeCmd.PersistentFlags().BoolVar(&doMergeAddOrigins, "merge-add-origins", false, "Add moz_origins to merge (formerly moz_hosts)")
	mergeCmd.PersistentFlags().IntVar(&limit, "limit", 0, "Limit the number of entries")
	RootCmd.AddCommand(mergeCmd)
}

func mergeRun(cmd *cobra.Command, args []string) {
	// check args
	if len(args) > 2 {
		Logger.Crit("merge has too many arguments")
	}
	if len(args) != 2 {
		Logger.Crit("merge requires your master places.sqlite and folder with places.sqlite files to merge")
	}
	if !FileExists(args[0]) {
		Logger.Critf("%s not found", args[0])
	}
	placesFolder, err := os.Stat(args[1])
	if err != nil {
		Logger.Critf("%s not found", args[1])
	}
	if !placesFolder.Mode().IsDir() {
		Logger.Critf("%s is not a directory", args[1])
	}

	// check and open db
	Logger.Printf("Checking and opening DBs...")
	masterPlacesDb, _, err = sqlite.OpenDbs(sqlite.SqliteFiles{
		Places: args[0], Favicons: "",
	}, true)
	if err != nil {
		Logger.Crit(err)
	}

	// working db infos
	Logger.Printf("\nMaster DB is '%s'", masterPlacesDb.Info.Filename)
	Logger.Debugf("Hash: %s:", masterPlacesDb.Info.Filehash)

	// backup db
	sqlite.BackupDb(masterPlacesDb.Info)

	// log
	mergeLogFile, err = os.OpenFile(PathJoin(appPath, "merge.log"), os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		Logger.Errorf("Merge log file: %s", err)
	}
	defer mergeLogFile.Close()
	mergeLog.Out = mergeLogFile
	mergeLog.Info("Starting merge...")

	// check merge flags
	mergeList := []string{"moz_places"}
	if doMergeFull {
		mergeList = append(mergeList, "moz_historyvisits (inc. in moz_places)", "moz_origins")
	} else {
		if doMergeAddHistoryvisits {
			mergeList = append(mergeList, "moz_historyvisits (inc. in moz_places)")
		}
		if doMergeAddOrigins {
			mergeList = append(mergeList, "moz_origins")
		}
	}
	Logger.Printf("\nThe following tables will be merged:\n- %s", strings.Join(mergeList, "\n- "))

	// open folder of places files to merge
	Logger.Printf("\nLooking for *.sqlite DBs in '%s'", args[1])
	slavePlacesDbs := sqlite.OpenPlacesDir(args[1], masterPlacesDb)
	Logger.Printf("%s valid DB(s) found:", strconv.Itoa(len(slavePlacesDbs)))
	for _, slavePlacesDb := range slavePlacesDbs {
		Logger.Printf("- %s (firefox %d v%d ; %d entries ; last used on %s)",
			slavePlacesDb.Info.Filename,
			slavePlacesDb.Info.FirefoxVersion,
			slavePlacesDb.Info.Version,
			slavePlacesDb.Info.PlacesCount,
			slavePlacesDb.Info.LastUsedTime.Format("2006-01-02 15:04:05"),
		)
	}

	// find max moz_places, moz_origins and moz_historyvisits ids
	lastID.MozPlaces = places.MozPlaces{}.GetLastID(masterPlacesDb.Link)
	lastID.MozOrigins = places.MozOrigins{}.GetLastID(masterPlacesDb.Link)
	lastID.MozHistoryvisits = places.MozHistoryvisits{}.GetLastID(masterPlacesDb.Link)
	Logger.Debugf("\nLast IDs found in working DB:")
	Logger.Debugf("- moz_places.id:        %d", lastID.MozPlaces)
	Logger.Debugf("- moz_origins.id:       %d", lastID.MozOrigins)
	Logger.Debugf("- moz_historyvisits.id: %d", lastID.MozHistoryvisits)

	// merge dbs to working db
	for _, slavePlacesDb := range slavePlacesDbs {
		Logger.Printf("\n## Merging DB '%s'...", slavePlacesDb.Info.Filename)
		merge(slavePlacesDb)
		slavePlacesDb.Link.Close()
	}

	Logger.Printf("\nOptimizing %s database...", masterPlacesDb.Info.Filename)
	if err = sqlite.Vacuum(masterPlacesDb.Link); err != nil {
		Logger.Warn(err)
	}

	masterPlacesDb.Link.Close()
}

func merge(slavePlacesDb sqlite.PlacesDb) {
	var placesResult MergeResult
	var historyvisitsResult MergeResult
	var originsResult MergeResult
	var hostsResult MergeResult

	// moz_places
	placesResult, historyvisitsResult = mergePlaces(slavePlacesDb)

	// moz_hosts or moz_origins
	if slavePlacesDb.Info.Version < 52 {
		hostsResult = mergeHosts(slavePlacesDb)
	} else {
		// TODO: check if places repaired with right origin_id
		originsResult = mergeOrigins(slavePlacesDb)
	}

	Logger.Debugf("\n[moz_places]")
	Logger.Debugf("  created = %d", placesResult.Created)
	Logger.Debugf("  updated = %d", placesResult.Updated)
	Logger.Debugf("  errors  = %d", placesResult.Errors)

	if historyvisitsResult.Created > 0 || historyvisitsResult.Errors > 0 {
		Logger.Debugf("\n[moz_historyvisits]")
		Logger.Debugf("  created = %d", historyvisitsResult.Created)
		Logger.Debugf("  errors  = %d", historyvisitsResult.Errors)
	}

	if originsResult.Created > 0 || originsResult.Updated > 0 || originsResult.Errors > 0 {
		Logger.Debugf("\n[moz_origins]")
		Logger.Debugf("  created = %d", originsResult.Created)
		Logger.Debugf("  updated = %d", originsResult.Updated)
		Logger.Debugf("  errors  = %d", originsResult.Errors)
	}

	if hostsResult.Created > 0 || hostsResult.Updated > 0 || hostsResult.Errors > 0 {
		Logger.Debugf("\n[moz_origins] from moz_hosts")
		Logger.Debugf("  created = %d", hostsResult.Created)
		Logger.Debugf("  updated = %d", hostsResult.Updated)
		Logger.Debugf("  errors  = %d", hostsResult.Errors)
	}

	if placesResult.Errors > 0 || historyvisitsResult.Errors > 0 || originsResult.Errors > 0 || hostsResult.Errors > 0 {
		cntErrors := placesResult.Errors + historyvisitsResult.Errors + originsResult.Errors + hostsResult.Errors
		Logger.Printf("\n%d error(s) occurred. Check your merge.log file", cntErrors)
	}
}

func mergePlaces(slavePlacesDb sqlite.PlacesDb) (placesResult MergeResult, historyvisitsResult MergeResult) {
	var slaveMozPlaces []places.MozPlaces
	if limit > 0 {
		slavePlacesDb.Link.Limit(limit).Find(&slaveMozPlaces)
	} else {
		slavePlacesDb.Link.Find(&slaveMozPlaces)
	}

	progBar := pb.StartNew(len(slaveMozPlaces))
	progBar.Prefix("moz_places")
	for _, slaveMozPlace := range slaveMozPlaces {
		var (
			masterMozPlace places.MozPlaces
			newMozPlace    places.MozPlaces
			updateMozPlace places.MozPlaces
		)
		masterPlacesDb.Link.Where("url = ?", slaveMozPlace.Url).Find(&masterMozPlace)

		// New moz_place entry
		if masterMozPlace.ID == 0 {
			lastID.MozPlaces++
			newMozPlace = slaveMozPlace
			newMozPlace.ID = lastID.MozPlaces
			newMozPlace.Guid = places.MozPlaces{}.GenerateGUID(slavePlacesDb.Link)
			if err := masterPlacesDb.Link.Create(&newMozPlace).Error; err != nil {
				mergeLog.Errorf("Creating moz_places row with id=%d : %s", newMozPlace.ID, err)
				placesResult.Errors++
			} else {
				placesResult.Created++
				mergeHistoryvisits(newMozPlace, slaveMozPlace, slavePlacesDb, &historyvisitsResult)
			}

			// Update moz_place entry
		} else {
			updateMozPlace = masterMozPlace
			updateMozPlace.VisitCount += slaveMozPlace.VisitCount
			updateMozPlace.LastVisitDate = MaxInt64(updateMozPlace.LastVisitDate, slaveMozPlace.LastVisitDate)
			updateMozPlace.Frecency = (updateMozPlace.Frecency + slaveMozPlace.Frecency) / 2
			if err := masterPlacesDb.Link.Save(&updateMozPlace).Error; err != nil {
				mergeLog.Errorf("Updating moz_places row with id=%d : %s", updateMozPlace.ID, err)
				placesResult.Errors++
			} else {
				placesResult.Updated++
				mergeHistoryvisits(updateMozPlace, slaveMozPlace, slavePlacesDb, &historyvisitsResult)
			}
		}
		progBar.Increment()
	}
	progBar.Finish()

	return placesResult, historyvisitsResult
}

func mergeHistoryvisits(masterMozPlaces places.MozPlaces, slaveMozPlaces places.MozPlaces, slavePlacesDb sqlite.PlacesDb, result *MergeResult) {
	if !doMergeFull && !doMergeAddHistoryvisits {
		return
	}

	var slaveMozHistoryvisits []places.MozHistoryvisits
	slavePlacesDb.Link.Where("place_id = ?", slaveMozPlaces.ID).Find(&slaveMozHistoryvisits)
	for _, slaveMozHistoryvisit := range slaveMozHistoryvisits {
		// Places already exists
		if masterMozPlaces.ID != lastID.MozPlaces {
			// Check history visit collisions before create
			var masterMozHistoryvisits places.MozHistoryvisits
			//FIXME: No unique columns so fingerprint with from_visit, place_id and visit_date
			masterPlacesDb.Link.Where("from_visit = ? AND place_id = ? AND visit_date = ?",
				slaveMozHistoryvisit.FromVisit,
				masterMozPlaces.ID,
				slaveMozHistoryvisit.VisitDate,
			).First(&masterMozHistoryvisits)

			if masterMozHistoryvisits.ID > 0 {
				return
			}
		}

		lastID.MozHistoryvisits++
		slaveMozHistoryvisit.ID = lastID.MozHistoryvisits
		slaveMozHistoryvisit.PlaceId = masterMozPlaces.ID
		slaveMozHistoryvisit.FromVisit = 0 //TODO: Find a way to retrieve ancestors. Fills from_visit with 0 temporarily.

		if err := masterPlacesDb.Link.Create(&slaveMozHistoryvisit).Error; err != nil {
			mergeLog.Errorf("Creating moz_historyvisits row with id=%d : %s", slaveMozHistoryvisit.ID, err)
			result.Errors++
		} else {
			result.Created++
		}
	}
}

func mergeHosts(slavePlacesDb sqlite.PlacesDb) MergeResult {
	var result MergeResult

	if !doMergeFull && !doMergeAddOrigins {
		return result
	}

	var slaveMozHosts []places.MozHosts
	slavePlacesDb.Link.Find(&slaveMozHosts)

	progBar := pb.StartNew(len(slaveMozHosts))
	progBar.Prefix("moz_hosts")
	for _, slaveMozHost := range slaveMozHosts {
		var masterMozOrigin places.MozOrigins
		masterPlacesDb.Link.Where("host = ?", slaveMozHost.Host).First(&masterMozOrigin)
		if masterMozOrigin.Host == "" {
			lastID.MozOrigins++
			masterMozOrigin.ID = lastID.MozOrigins
			masterMozOrigin.Prefix = slaveMozHost.Prefix
			masterMozOrigin.Host = slaveMozHost.Host
			masterMozOrigin.Frecency = slaveMozHost.Frecency
			if err := masterPlacesDb.Link.Create(&masterMozOrigin).Error; err != nil {
				mergeLog.Errorf("Creating moz_origins row with id=%d : %s", masterMozOrigin.ID, err)
				result.Errors++
			} else {
				result.Created++
			}
		} else {
			masterMozOrigin.Frecency = (masterMozOrigin.Frecency + slaveMozHost.Frecency) / 2
			if err := masterPlacesDb.Link.Save(&masterMozOrigin).Error; err != nil {
				mergeLog.Errorf("Updating moz_origins row with id=%d : %s", masterMozOrigin.ID, err)
				result.Errors++
			} else {
				result.Updated++
			}
		}
		progBar.Increment()
	}
	progBar.Finish()

	return result
}

func mergeOrigins(slavePlacesDb sqlite.PlacesDb) MergeResult {
	var result MergeResult

	if !doMergeFull && !doMergeAddOrigins {
		return result
	}

	var slaveMozOrigins []places.MozOrigins
	slavePlacesDb.Link.Find(&slaveMozOrigins)

	progBar := pb.StartNew(len(slaveMozOrigins))
	progBar.Prefix("moz_origins")
	for _, slaveMozOrigin := range slaveMozOrigins {
		var masterMozOrigin places.MozOrigins
		masterPlacesDb.Link.Where("host = ?", slaveMozOrigin.Host).First(&masterMozOrigin)
		if masterMozOrigin.Host == "" {
			lastID.MozOrigins++
			slaveMozOrigin.ID = lastID.MozOrigins
			if err := masterPlacesDb.Link.Create(&slaveMozOrigin).Error; err != nil {
				mergeLog.Errorf("Creating moz_origins row with id=%d : %s", slaveMozOrigin.ID, err)
				result.Errors++
			} else {
				result.Created++
			}
		} else {
			masterMozOrigin.Frecency = (masterMozOrigin.Frecency + slaveMozOrigin.Frecency) / 2
			if err := masterPlacesDb.Link.Save(&masterMozOrigin).Error; err != nil {
				mergeLog.Errorf("Updating moz_origins row with id=%d : %s", masterMozOrigin.ID, err)
				result.Errors++
			} else {
				result.Updated++
			}
		}
		progBar.Increment()
	}
	progBar.Finish()

	return result
}
